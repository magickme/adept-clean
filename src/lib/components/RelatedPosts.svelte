<script lang="ts">
    let { relatedPosts } = $props<{ relatedPosts: any[] }>();

    function formatDate(date: string): string {
        return new Date(date).toLocaleDateString('en-US', {
            year: 'numeric',
            month: 'long',
            day: 'numeric'
        });
    }

    function truncateText(text: string = '', maxLength: number): string {
        return text?.length > maxLength ? text.slice(0, maxLength) + '...' : text || '';
    }

    function getAuthorName(post: any): string {
        return post?.author?.name || '';
    }

    function getAuthorImage(post: any): string {
        return post?.author?.profile_image || '';
    }

    function getReadingTime(post: any): string {
        return post?.reading_time ? `${post.reading_time} min read` : '';
    }
</script>

<div data-testid="RelatedPosts" class="bg-kohl py-24 sm:py-32">
    <div class="mx-auto max-w-7xl px-6 lg:px-8">
        <div class="mx-auto max-w-2xl text-center">
            <h2 class="text-h2 text-ivory font-montserrat">
                Browse related articles
            </h2>
            <p class="mt-2 text-p text-flannel font-sourceSans">
                Learn more about our magical journey and insights.
            </p>
        </div>
        <div class="mx-auto mt-16 grid max-w-2xl grid-cols-1 gap-x-8 gap-y-20 lg:mx-0 lg:max-w-none lg:grid-cols-3">
            {#each relatedPosts as post}
                <article class="flex flex-col items-start">
                    <div class="relative w-full">
                        <img
                            src={post.feature_image || ''}
                            alt={post.title || ''}
                            class="aspect-[16/9] w-full rounded-2xl bg-flannel/10 object-cover sm:aspect-[2/1] lg:aspect-[3/2]"
                        />
                        <div class="absolute inset-0 rounded-2xl ring-1 ring-inset ring-flannel/10"></div>
                    </div>
                    <div class="max-w-xl">
                        <div class="mt-8 flex items-center gap-x-4 text-cap">
                            <time datetime={post.published_at} class="text-flannel font-sourceSans">
                                {formatDate(post.published_at)}
                            </time>
                            {#each post.tags as tag}
                                <a
                                    href={`/blog/tag/${tag}`}
                                    class="relative z-10 rounded-full bg-cerulean px-3 py-1.5 text-ivory hover:bg-corsair font-montserrat"
                                >
                                    {tag}
                                </a>
                            {/each}
                        </div>
                        <div class="group relative">
                            <h3 class="mt-3 text-h4 text-ivory group-hover:text-cerulean font-montserrat">
                                <a href={`/blog/${post.slug || ''}`}>
                                    <span class="absolute inset-0"></span>
                                    {post.title || ''}
                                </a>
                            </h3>
                            <p class="mt-5 text-p text-flannel font-sourceSans">
                                {truncateText(post.excerpt, 100)}
                            </p>
                        </div>
                        <div class="relative mt-8 flex items-center gap-x-4">
                            <img 
                                src={getAuthorImage(post)} 
                                alt="" 
                                class="h-10 w-10 rounded-full bg-flannel/10"
                                width="40"
                                height="40" 
                            />
                            <div class="text-p leading-6">
                                <p class="font-montserrat text-ivory">
                                    <a href={`/blog/author/${getAuthorName(post)}`}>
                                        <span class="absolute inset-0"></span>
                                        {getAuthorName(post)}
                                    </a>
                                </p>
                                <p class="text-flannel font-sourceSans">
                                    {getReadingTime(post)}
                                </p>
                            </div>
                        </div>
                    </div>
                </article>
            {/each}
        </div>
    </div>
</div>