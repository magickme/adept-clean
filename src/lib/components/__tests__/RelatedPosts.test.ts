import { render, screen } from '@testing-library/svelte/svelte5';
import { expect, describe, it } from 'vitest';
import RelatedPosts from '../RelatedPosts.svelte';
import type { Post } from '$lib/types';

describe('RelatedPosts', () => {
  // Constants
  const SELECTORS = {
    excerpt: '.text-gray-500.dark\\:text-gray-300',
    container: '[data-testid="RelatedPosts"]'
  } as const;

  const TEST_IDS = {
    container: 'RelatedPosts'
  } as const;

  const TEXTS = {
    heading: 'Browse related articles',
    readingTime: 'min read'
  } as const;

  const CLASSES = {
    darkMode: {
      container: 'dark:bg-kohl',
      article: 'dark:bg-gray-800'
    }
  } as const;

  const PERFORMANCE = {
    renderThreshold: 100, // ms
    maxPosts: 50
  } as const;

  // Types
  interface SetupResult {
    container: HTMLElement;
    getContainer: () => HTMLElement;
    getArticles: () => HTMLElement[];
    getTimeElements: () => NodeListOf<HTMLTimeElement>;
    getFeatureImage: (article: HTMLElement) => HTMLImageElement | null;
    getTitleLink: (title: string) => HTMLAnchorElement | null;
    getExcerptElement: (article: HTMLElement) => HTMLParagraphElement | null;
  }

  // Test data factory
  const createMockPost = (overrides: Partial<Post> = {}): Post => ({
    title: 'Test Post 1',
    slug: 'test-post-1',
    feature_image: '/test1.jpg',
    published_at: '2024-01-01T00:00:00.000Z',
    excerpt: 'Test excerpt 1',
    author: {
      name: 'John Doe',
      profile_image: '/author1.jpg'
    },
    reading_time: 5,
    tags: ['tag1', 'tag2'],
    ...overrides
  });

  const createMockPosts = (count = 2): Post[] => 
    Array(count).fill(null).map((_, index) => 
      createMockPost({
        title: `Test Post ${index + 1}`,
        slug: `test-post-${index + 1}`,
        feature_image: `/test${index + 1}.jpg`,
        published_at: new Date(2024, 0, index + 1).toISOString(),
        author: {
          name: index === 0 ? 'John Doe' : 'Jane Smith',
          profile_image: `/author${index + 1}.jpg`
        },
        reading_time: 5 * (index + 1),
        tags: index === 0 ? ['tag1', 'tag2'] : ['tag3']
      })
    );

  // Helper functions
  const setup = (posts: Post[] = createMockPosts()): SetupResult => {
    const utils = render(RelatedPosts, { relatedPosts: posts });
    
    return {
      ...utils,
      getContainer: () => screen.getByTestId(TEST_IDS.container),
      getArticles: () => Array.from(screen.getAllByRole('article')),
      getTimeElements: () => utils.container.querySelectorAll('time'),
      getFeatureImage: (article: HTMLElement) => article.querySelector('img'),
      getTitleLink: (title: string) => screen.getByText(title).closest('a'),
      getExcerptElement: (article: HTMLElement) => 
        article.querySelector<HTMLParagraphElement>(SELECTORS.excerpt)
    };
  };

  // Test groups
  describe('rendering', () => {
    it('renders the component with correct structure', () => {
      const { getContainer, getArticles } = setup();
      
      expect(getContainer()).toBeInTheDocument();
      expect(screen.getByText(TEXTS.heading)).toBeInTheDocument();
      expect(getArticles()).toHaveLength(2);
    });

    it('handles empty posts array gracefully', () => {
      const { getContainer } = setup([]);
      
      expect(getContainer()).toBeInTheDocument();
      expect(screen.queryByRole('article')).not.toBeInTheDocument();
    });

    it('matches snapshot', () => {
      const { container } = setup();
      // Clean up Svelte-generated comments before snapshot
      const cleanContainer = container.cloneNode(true) as HTMLElement;
      Array.from(cleanContainer.querySelectorAll('*'))
        .forEach(element => {
          // Remove comment nodes
          element.childNodes.forEach(node => {
            if (node.nodeType === Node.COMMENT_NODE) {
              node.remove();
            }
          });
        });
      
      expect(cleanContainer).toMatchSnapshot();
    });
  });

  describe('content', () => {
    describe('post details', () => {
      it('renders post content correctly', () => {
        const mockPost = createMockPost();
        const { getArticles, getFeatureImage, getTitleLink } = setup([mockPost]);
        
        const article = getArticles()[0];
        
        // Image
        const featureImage = getFeatureImage(article);
        expect(featureImage).toHaveAttribute('src', mockPost.feature_image);
        
        // Title and link
        const titleLink = getTitleLink(mockPost.title);
        expect(titleLink).toHaveAttribute('href', `/blog/${mockPost.slug}`);
        
        // Date
        const timeElement = article.querySelector('time');
        expect(timeElement).toHaveAttribute('datetime', mockPost.published_at);
        
        // Author
        expect(screen.getByText(mockPost.author.name)).toBeInTheDocument();
        const authorImage = screen.getByAltText(mockPost.author.name);
        expect(authorImage).toHaveAttribute('src', mockPost.author.profile_image);
        
        // Reading time
        expect(screen.getByText(`${mockPost.reading_time} ${TEXTS.readingTime}`)).toBeInTheDocument();
        
        // Tags
        mockPost.tags.forEach(tag => {
          expect(screen.getByText(tag)).toBeInTheDocument();
        });
      });

      it('handles malformed post data gracefully', () => {
        const malformedPost = createMockPost({ 
          author: undefined,
          tags: undefined,
          feature_image: undefined,
          reading_time: undefined
        } as any);
        
        const { getArticles, getFeatureImage } = setup([malformedPost]);
        const article = getArticles()[0];
        
        expect(article).toBeInTheDocument();
        expect(getFeatureImage(article)).toHaveAttribute('alt', '');
        expect(screen.queryByText(`${malformedPost.reading_time} ${TEXTS.readingTime}`))
          .not.toBeInTheDocument();
      });

      describe('error handling', () => {
        it('handles missing author data gracefully', () => {
          const postWithoutAuthor = createMockPost({ 
            author: {
              name: '',
              profile_image: ''
            }
          });
          
          const { getArticles } = setup([postWithoutAuthor]);
          const article = getArticles()[0];
          
          expect(article).toBeInTheDocument();
          const authorImage = article.querySelector('.h-10.w-10.rounded-full');
          expect(authorImage).toHaveAttribute('src', '');
          expect(authorImage).toHaveAttribute('alt', '');
        });

        it('handles completely malformed post data gracefully', () => {
          const malformedPost = {
            title: '',
            slug: '',
            feature_image: '',
            published_at: '',
            excerpt: '',
            author: null,
            reading_time: null,
            tags: null
          } as unknown as Post;
          
          const { getArticles, getFeatureImage } = setup([malformedPost]);
          const article = getArticles()[0];
          
          expect(article).toBeInTheDocument();
          expect(getFeatureImage(article)).toHaveAttribute('alt', '');
          expect(screen.queryByText('null min read')).not.toBeInTheDocument();
        });

        it('handles undefined post properties gracefully', () => {
          const postWithUndefinedProps = createMockPost({
            feature_image: undefined,
            excerpt: undefined,
            reading_time: undefined,
            tags: undefined,
            author: undefined
          } as Partial<Post>);
          
          const { getArticles } = setup([postWithUndefinedProps]);
          expect(getArticles()).toHaveLength(1);
        });
      });
    });

    describe('text handling', () => {
      it('truncates long excerpts correctly', () => {
        const longText = 'a'.repeat(150);
        const mockPost = createMockPost({ excerpt: longText });
        const { getArticles, getExcerptElement } = setup([mockPost]);
        
        const article = getArticles()[0];
        const excerptElement = getExcerptElement(article);
        
        expect(excerptElement).toBeInTheDocument();
        expect(excerptElement?.textContent?.length).toBeLessThanOrEqual(103);
        expect(excerptElement?.textContent?.endsWith('...')).toBe(true);
      });
    });
  });

  describe('accessibility', () => {
    it('meets accessibility requirements', () => {
      setup();
      
      // Images have alt text
      const images = screen.getAllByRole('img');
      images.forEach(img => {
        expect(img).toHaveAttribute('alt');
      });

      // Links are accessible
      const links = screen.getAllByRole('link');
      links.forEach(link => {
        expect(link).toHaveAttribute('href');
      });

      // Heading hierarchy is correct
      const mainHeading = screen.getByRole('heading', { level: 2 });
      expect(mainHeading).toHaveTextContent(TEXTS.heading);
    });
  });

  describe('dates', () => {
    it('formats dates correctly', () => {
      const mockPosts = createMockPosts();
      const { getTimeElements } = setup(mockPosts);
      
      const timeElements = getTimeElements();
      expect(timeElements[0]).toHaveAttribute('datetime', mockPosts[0].published_at);
      expect(timeElements[1]).toHaveAttribute('datetime', mockPosts[1].published_at);
    });
  });

  describe('styling', () => {
    it('includes dark mode classes', () => {
      const { getContainer, getArticles } = setup();
      
      expect(getContainer()).toHaveClass(CLASSES.darkMode.container);
      
      getArticles().forEach(article => {
        expect(article).toHaveClass(CLASSES.darkMode.article);
      });
    });
  });

  describe('performance', () => {
    it('renders efficiently with many posts', () => {
      const manyPosts = createMockPosts(PERFORMANCE.maxPosts);
      
      const start = performance.now();
      const { getArticles } = setup(manyPosts);
      const duration = performance.now() - start;
      
      expect(duration).toBeLessThan(PERFORMANCE.renderThreshold);
      expect(getArticles()).toHaveLength(PERFORMANCE.maxPosts);
    });
  });
}); 