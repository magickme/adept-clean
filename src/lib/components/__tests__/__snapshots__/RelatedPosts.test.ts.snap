// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RelatedPosts > rendering > matches snapshot 1`] = `
<div>
  <div
    class="bg-ivory dark:bg-kohl py-12 sm:py-16"
    data-testid="RelatedPosts"
  >
    <div
      class="mx-auto max-w-7xl px-6 lg:px-8"
    >
      <div
        class="mx-auto max-w-2xl text-center"
      >
        <h2
          class="text-3xl font-bold tracking-tight text-gray-900 dark:text-gray-100 sm:text-4xl"
        >
          Browse related articles
        </h2>
      </div>
       
      <div
        class="mx-auto mt-16 grid max-w-2xl auto-rows-fr grid-cols-1 gap-8 sm:mt-20 lg:mx-0 lg:max-w-none lg:grid-cols-3"
      >
        <article
          class="relative flex flex-col overflow-hidden rounded-lg bg-white dark:bg-gray-800 shadow-lg"
        >
          <div
            class="h-48 w-full overflow-hidden"
          >
            <img
              alt=""
              class="h-full w-full object-cover"
              src="/test1.jpg"
            />
          </div>
           
          <div
            class="flex flex-1 flex-col justify-between p-6"
          >
            <div
              class="flex-1"
            >
              <p
                class="text-sm font-medium text-cerulean"
              >
                <time
                  datetime="2024-01-01T06:00:00.000Z"
                >
                  January 1, 2024
                </time>
              </p>
               
              <a
                class="mt-2 block"
                href="/blog/test-post-1"
              >
                <h3
                  class="text-xl font-semibold text-gray-900 dark:text-white"
                >
                  Test Post 1
                </h3>
                 
                <p
                  class="mt-3 text-base text-gray-500 dark:text-gray-300"
                >
                  Test excerpt 1
                </p>
              </a>
            </div>
             
            <div
              class="mt-6 flex items-center"
            >
              <div
                class="flex-shrink-0"
              >
                <img
                  alt="John Doe"
                  class="h-10 w-10 rounded-full"
                  src="/author1.jpg"
                />
              </div>
               
              <div
                class="ml-3"
              >
                <p
                  class="text-sm font-medium text-gray-900 dark:text-white"
                >
                  John Doe
                </p>
                 
                <div
                  class="flex space-x-1 text-sm text-gray-500 dark:text-gray-400"
                >
                  <span>
                    5 min read
                  </span>
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="absolute bottom-0 right-0 p-2"
          >
            <span
              class="inline-block bg-gray-200 dark:bg-gray-700 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 dark:text-gray-300 mr-2 mb-2"
            >
              tag1
            </span>
            <span
              class="inline-block bg-gray-200 dark:bg-gray-700 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 dark:text-gray-300 mr-2 mb-2"
            >
              tag2
            </span>
            
          </div>
        </article>
        <article
          class="relative flex flex-col overflow-hidden rounded-lg bg-white dark:bg-gray-800 shadow-lg"
        >
          <div
            class="h-48 w-full overflow-hidden"
          >
            <img
              alt=""
              class="h-full w-full object-cover"
              src="/test2.jpg"
            />
          </div>
           
          <div
            class="flex flex-1 flex-col justify-between p-6"
          >
            <div
              class="flex-1"
            >
              <p
                class="text-sm font-medium text-cerulean"
              >
                <time
                  datetime="2024-01-02T06:00:00.000Z"
                >
                  January 2, 2024
                </time>
              </p>
               
              <a
                class="mt-2 block"
                href="/blog/test-post-2"
              >
                <h3
                  class="text-xl font-semibold text-gray-900 dark:text-white"
                >
                  Test Post 2
                </h3>
                 
                <p
                  class="mt-3 text-base text-gray-500 dark:text-gray-300"
                >
                  Test excerpt 1
                </p>
              </a>
            </div>
             
            <div
              class="mt-6 flex items-center"
            >
              <div
                class="flex-shrink-0"
              >
                <img
                  alt="Jane Smith"
                  class="h-10 w-10 rounded-full"
                  src="/author2.jpg"
                />
              </div>
               
              <div
                class="ml-3"
              >
                <p
                  class="text-sm font-medium text-gray-900 dark:text-white"
                >
                  Jane Smith
                </p>
                 
                <div
                  class="flex space-x-1 text-sm text-gray-500 dark:text-gray-400"
                >
                  <span>
                    10 min read
                  </span>
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="absolute bottom-0 right-0 p-2"
          >
            <span
              class="inline-block bg-gray-200 dark:bg-gray-700 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 dark:text-gray-300 mr-2 mb-2"
            >
              tag3
            </span>
            
          </div>
        </article>
        
      </div>
    </div>
  </div>
  
</div>
`;
